### ISSUE TITLE
FP `ptr-cast-constness`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#11317]
### Summary

.

### Lint Name

ptr-cast-constness

### Reproducer

I tried this code:

```rust
#![allow(invalid_reference_casting)]

fn main() {
    let target = 42;
    // Make sure raw ptr with raw tag cannot mutate frozen location without breaking the shared ref.
    let r#ref = &target; // freeze
    let ptr = r#ref as *const _ as *mut _; // raw ptr, with raw tag
    unsafe { *ptr = 42 }; //~ ERROR: /write access .* only grants SharedReadOnly permission/
    let _val = *r#ref;
}

```

I saw this happen:

```
warning: `as` casting between raw pointers while changing only its constness
 --> src/main.rs:7:15
  |
7 |     let ptr = r#ref as *const _ as *mut _; // raw ptr, with raw tag
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try `pointer::cast_mut`, a safer alternative: `(r#ref as *const _).cast_mut()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_cast_constness
  = note: requested on the command line with `-W clippy::ptr-cast-constness`
```
rustfix error:

````
The following errors were reported:
error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
 --> src/main.rs:7:35
  |
7 |     let ptr = (r#ref as *const _).cast_mut(); // raw ptr, with raw tag
  |                                   ^^^^^^^^

error: aborting due to previous error
````

### Version

```text
rustc 1.73.0-nightly (f88a8b71c 2023-08-08)
binary: rustc
commit-hash: f88a8b71cebb730cbd5058c45ebcae1d4d9be377
commit-date: 2023-08-08
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 17.0.0
```

**Comment by asquared31415** [comment]
minimized, and with an attribute to enable the lint, since that's easier to test
```rust
#![warn(clippy::ptr_cast_constness)]

fn main() {
    let r = &0_u32;
    let _ptr: *mut u32 = r as *const _ as *mut _;
}
```

**Comment by Centri3** [comment]
We need to check for an `Infer` (`_`) `TyKind`. If the type of the pointer is that, it shouldn't lint (or suggest changing some other part of the code, if `ptr_as_ptr` is enabled)

Don't recall whether the typeck table contains it though, if not we'll need to be a bit more clever.
